packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart

```diff
@@ -1,17 +1,349 @@
+                child: _flameChartBody(constraints),
+              ),
+            ),
+          ),
+        );
+      },
+    );
+  }
+
+  Widget _flameChartBody(BoxConstraints constraints) {
+    final width = math.max(constraints.maxWidth, widget.totalStartingWidth);
+    final height = math.max(constraints.maxHeight, widget.height);
+
+    return Stack(
+      children: [
+        Container(
+          width: width,
+          height: height,
+        ),
+        ..._nodesInViewport(constraints), // pick what to show
+      ],
+    );
+  }
+
+  List<FlameChartNode> _nodesInViewport(BoxConstraints constraints) {
+    // TODO(kenz): is creating all the FlameChartNode objects expensive even if
+    // we won't add them to the view? We create all the FlameChartNode objects
+    // and place them in FlameChart rows, but we only add [nodesInViewport] to
+    // the widget tree.
+    _buildFlameChartElements();
+
+    final nodesInViewport = <FlameChartNode>[];
+    for (var row in rows) {
+      for (var node in row.nodes) {
+        final fitsHorizontally = node.rect.right >= scrollOffsetX &&
+            node.rect.left - scrollOffsetX <= constraints.maxWidth;
+        final fitsVertically = node.rect.bottom >= scrollOffsetY &&
+            node.rect.top - scrollOffsetY <= constraints.maxHeight;
+        if (fitsHorizontally && fitsVertically) {
+          nodesInViewport.add(node);
+        }
+      }
+    }
+    return nodesInViewport;
+  }
+
+  void _buildFlameChartElements() {
+    _resetColorOffsets();
+
+    rows
+      ..clear()
+      ..addAll(List.generate(
+        widget.data.uiEventFlow.depth + widget.data.gpuEventFlow.depth,
+        (i) => FlameChartRow(nodes: [], index: i),
+      ));
+    final int frameStartOffset = widget.data.time.start.inMicroseconds;
+
+    double getTopForRow(int row) {
+      // This accounts for the section spacing between the UI events and the GPU
+      // events.
+      final additionalPadding =
+          row >= gpuSectionStartRow ? sectionSpacing : 0.0;
+      return (row * rowHeightWithPadding + topOffset + additionalPadding)
+          .toDouble();
+    }
+
+    // Pixels per microsecond in order to fit the entire frame in view.
+    final double pxPerMicro =
+        widget.startingContentWidth / widget.data.time.duration.inMicroseconds;
+
+    // Add UI section label.
+    final uiSectionLabel = FlameChartNode.sectionLabel(
+      text: 'UI',
+      textColor: Colors.black,
+      backgroundColor: mainUiColor,
+      top: getTopForRow(0),
+      width: 28.0,
+    );
+    rows[0].nodes.add(uiSectionLabel);
+
+    // Add GPU section label.
+    final gpuSectionLabel = FlameChartNode.sectionLabel(
+      text: 'GPU',
+      textColor: Colors.white,
+      backgroundColor: mainGpuColor,
+      top: getTopForRow(gpuSectionStartRow),
+      width: 42.0,
     );
+    rows[gpuSectionStartRow].nodes.add(gpuSectionLabel);
+
+    void createChartNodes(TimelineEvent event, int row) {
```

Comment:
Is it possible to pass in the bounds of the viewport here and use this to determine which events from the row to create a node for?

Reply:
Will leave this as is for now, as discussed offline. I'll add a TODO to track the suggestion for when we try to optimize this code later.